File: BiBoGC/Sale.Domain/Domain/Invoice.cs
using Sale.Domain.Enum;
using Sale.Domain.Events;
using Sale.Domain.ValueObjects;
using Shared.Domain.Common;

namespace Sale.Domain.Domain;

/// <summary>
/// Hóa đơn - Immutable sau khi tạo
/// </summary>
public class Invoice : BaseEntity
{
    private readonly List<InvoiceItem> _items = [];

    private Invoice()
    {
    }

    private Invoice(
        string invoiceNumber,
        Guid salesOrderId,
        string orderNumber,
        StoreInfo storeInfo,
        string? customerName,
        string? customerPhone,
        decimal subTotal,
        decimal discountAmount,
        decimal taxAmount,
        decimal grandTotal,
        decimal amountPaid,
        string paymentMethod,
        IEnumerable<InvoiceItem> items)
    {
        InvoiceNumber = invoiceNumber;
        InvoiceDate = DateTime.UtcNow;
        SalesOrderId = salesOrderId;
        OrderNumber = orderNumber;

        StoreName = storeInfo.Name;
        StoreAddress = storeInfo.Address;
        StorePhone = storeInfo.Phone;
        StoreTaxCode = storeInfo.TaxNumber;

        CustomerName = customerName;
        CustomerPhone = customerPhone;

        SubTotal = subTotal;
        DiscountAmount = discountAmount;
        TaxAmount = taxAmount;
        GrandTotal = grandTotal;
        AmountPaid = amountPaid;
        ChangeAmount = amountPaid - grandTotal;
        PaymentMethod = paymentMethod;

        foreach (var item in items)
        {
            _items.Add(item);
        }

        RaiseDomainEvent(new InvoiceGeneratedEvent(Id, InvoiceNumber, SalesOrderId, OrderNumber));
    }

    #region Factory Method

    /// <summary>
    /// Tạo Invoice từ SalesOrder đã hoàn thành
    /// </summary>
    public static Invoice CreateFromOrder(
        SalesOrder order,
        StoreInfo storeInfo,
        string invoiceNumber)
    {
        if (order.Status != OrderStatus.Completed)
            throw new InvalidOperationException("Chỉ có thể tạo hóa đơn cho đơn hàng đã hoàn thành.");

        var invoiceItems = order.Items.Select(item => new InvoiceItem(
            productName: item.ProductName,
            variantName: item.VariantName,
            sku: item.Sku,
            unit: item.Unit,
            quantity: item.Quantity,
            unitPrice: item.UnitPrice,
            lineTotal: item.LineTotal
        )).ToList();

        var invoice = new Invoice(
            invoiceNumber: invoiceNumber,
            salesOrderId: order.Id,
            orderNumber: order.OrderNumber,
            storeInfo: storeInfo,
            customerName: order.CustomerName,
            customerPhone: order.CustomerPhone,
            subTotal: order.SubTotal,
            discountAmount: order.DiscountAmount,
            taxAmount: order.TaxAmount,
            grandTotal: order.TotalAmount,
            amountPaid: order.AmountPaid,
            paymentMethod: order.PaymentMethod.ToString(),
            items: invoiceItems
        );

        foreach (var item in invoice._items)
        {
            item.SetInvoiceId(invoice.Id);
        }

        return invoice;
    }

    #endregion

    #region Properties

    public string InvoiceNumber { get; private set; } = null!;
    public DateTime InvoiceDate { get; private set; }
    public Guid SalesOrderId { get; private set; }
    public string OrderNumber { get; private set; } = null!;

    public string StoreName { get; private set; } = null!;
    public string StoreAddress { get; private set; } = null!;
    public string StorePhone { get; private set; } = null!;
    public string? StoreTaxCode { get; private set; }

    public string? CustomerName { get; private set; }
    public string? CustomerPhone { get; private set; }

    public decimal SubTotal { get; private set; }
    public decimal DiscountAmount { get; private set; }
    public decimal TaxAmount { get; private set; }
    public decimal GrandTotal { get; private set; }
    public decimal AmountPaid { get; private set; }
    public decimal ChangeAmount { get; private set; }
    public string PaymentMethod { get; private set; } = null!;

    public IReadOnlyCollection<InvoiceItem> Items => _items.AsReadOnly();

    #endregion
}

File: FrontEnd/bibogc/lib/features/invoice/domain/entities/invoice.dart
class Invoice {
  final String id;
  final String invoiceNumber;
  final DateTime invoiceDate;
  final String salesOrderId;
  final String orderNumber;
  final String storeName;
  final String storeAddress;
  final String storePhone;
  final String? storeTaxCode;
  final String? customerName;
  final String? customerPhone;
  final double subTotal;
  final double discountAmount;
  final double taxAmount;
  final double grandTotal;
  final double amountPaid;
  final double changeAmount;
  final String paymentMethod;
  final List<InvoiceItem> items;
  final DateTime createdAt;

  const Invoice({
    required this.id,
    required this.invoiceNumber,
    required this.invoiceDate,
    required this.salesOrderId,
    required this.orderNumber,
    required this.storeName,
    required this.storeAddress,
    required this.storePhone,
    this.storeTaxCode,
    this.customerName,
    this.customerPhone,
    required this.subTotal,
    required this.discountAmount,
    required this.taxAmount,
    required this.grandTotal,
    required this.amountPaid,
    required this.changeAmount,
    required this.paymentMethod,
    required this.items,
    required this.createdAt,
  });

  String get paymentMethodLabel => switch (paymentMethod) {
    'Cash' => 'Tiền mặt',
    'BankTransfer' => 'Chuyển khoản',
    _ => paymentMethod,
  };
}

class InvoiceItem {
  final String id;
  final String productName;
  final String variantName;
  final String sku;
  final String unit;
  final int quantity;
  final double unitPrice;
  final double lineTotal;

  const InvoiceItem({
    required this.id,
    required this.productName,
    required this.variantName,
    required this.sku,
    required this.unit,
    required this.quantity,
    required this.unitPrice,
    required this.lineTotal,
  });
}
